core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,287 @@
+import java.util.zip.ZipOutputStream;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+
+/**
+ * XMLRecordReader class to read through a given xml document and to output xml blocks as per node path specified.
+ */
+public class XMLRecordReader extends RecordReader<LongWritable, Map<String, String>> {
+
+  public static final String CLOSING_END_TAG_DELIMITER = ">";
+  public static final String OPENING_END_TAG_DELIMITER = "</";
+  public static final String CLOSING_START_TAG_DELIMITER = ">";
+  public static final String OPENING_START_TAG_DELIMITER = "<";
+
+  private LongWritable currentKey;
+  private Map<String, String> currentValue;
+  private String nodePath;
+  private String fileName;
+  private XMLStreamReader reader;
+  private String[] nodes;
+  private int totalNodes;
+  private Map<Integer, String> actualNodeLevelMap;
+  private Map<Integer, String> currentNodeLevelMap;
+  private int nodeLevel = 0;
+  private String tempFilePath = null;
+  private Path file;
+  private String fileAction;
+  private FileSystem fs;
+  private String targetFolder;
+
+  public XMLRecordReader(FileSplit split, Configuration conf) throws IOException {
+      file = split.getPath();
+      fileName = file.toUri().toString();
+      fs = file.getFileSystem(conf);
+      XMLInputFactory factory = XMLInputFactory.newInstance();
+      FSDataInputStream fdDataInputStream = fs.open(file);
+      try {
+      reader = factory.createXMLStreamReader(fdDataInputStream);
+      } catch (XMLStreamException exception) {
+        throw new RuntimeException("XMLStreamException exception : ", exception);
+      }
+      //set required node path details.
+      nodePath = conf.get(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH);
+      //remove preceding '/' in node path to avoid first unwanted element after split('/')
+      if (nodePath.indexOf("/") == 0) {
+        nodePath = nodePath.substring(1, nodePath.length());
+      }
+      nodes = nodePath.split("/");
+      totalNodes = nodes.length;
+
+      //map to store node path specified nodes with key as node level
+      actualNodeLevelMap = new HashMap<Integer, String>();
+      for (int i = 0; i < totalNodes; i++) {
+        actualNodeLevelMap.put(i, nodes[i]);
+      }
+      currentNodeLevelMap = new HashMap<Integer, String>();
+
+      tempFilePath = conf.get(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FILE);
+      fileAction = conf.get(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION);
+      targetFolder = conf.get(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (reader != null) {
+      try {
+        reader.close();
+      } catch (XMLStreamException exception) {
+        // Swallow exception.
+      }
+    }
+  }
+
+  @Override
+  public float getProgress() throws IOException {
+    return 1.0f;
```

Comment:
Difficult to determine the progress with streaming. I have tried using FilterInputStream, see this **[Gist](https://gist.github.com/bhushankawadkar/52f5ae4a86724d5710e763f02dd89b6e)**, but this does not guarantees the correct result. Please provide your inputs.